fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/properties/NormalizedResidualPredicateProperty.java

```diff
@@ -49,64 +49,83 @@
+    private NormalizedResidualPredicateProperty() {
+        // prevent outside instantiation
     }
 
     @Nonnull
     @Override
-    public QueryPredicate evaluateAtExpression(@Nonnull RelationalExpression expression, @Nonnull List<QueryPredicate> childResults) {
-        final var resultPredicatesBuilder = ImmutableList.<QueryPredicate>builder();
-
-        childResults.stream()
-                .filter(Objects::nonNull)
-                .filter(predicate -> !predicate.isTautology())
-                .forEach(resultPredicatesBuilder::add);
-
-        if (expression instanceof RelationalExpressionWithPredicates) {
-            ((RelationalExpressionWithPredicates)expression).getPredicates()
-                    .stream()
-                    .filter(predicate -> !predicate.isTautology())
-                    .forEach(resultPredicatesBuilder::add);
-        }
-
-        return AndPredicate.and(resultPredicatesBuilder.build());
+    public NormalizedResidualPredicateVisitor createVisitor() {
+        return new NormalizedResidualPredicateVisitor();
     }
 
     @Nonnull
-    @Override
-    public QueryPredicate evaluateAtRef(@Nonnull Reference ref, @Nonnull List<QueryPredicate> memberResults) {
-        Verify.verify(memberResults.size() == 1);
-        return Iterables.getOnlyElement(memberResults);
+    public QueryPredicate evaluate(@Nonnull final Reference reference) {
+        return Objects.requireNonNull(reference.acceptPropertyVisitor(createVisitor()));
     }
 
     @Nonnull
-    public static QueryPredicate evaluate(Reference ref) {
-        return Verify.verifyNotNull(ref.acceptPropertyVisitor(INSTANCE));
+    public QueryPredicate evaluate(@Nonnull final RelationalExpression expression) {
+        return Objects.requireNonNull(expression.acceptPropertyVisitor(createVisitor()));
     }
 
     @Nonnull
-    public static QueryPredicate evaluate(@Nonnull RelationalExpression expression) {
-        return Verify.verifyNotNull(expression.acceptPropertyVisitor(INSTANCE));
+    public static NormalizedResidualPredicateProperty normalizedResidualPredicate() {
+        return NORMALIZED_RESIDUAL_PREDICATE;
     }
 
     public static long countNormalizedConjuncts(@Nonnull RelationalExpression expression) {
-        final var magicPredicate = evaluate(expression);
+        final var magicPredicate = normalizedResidualPredicate().evaluate(expression);
         return magicPredicate.isTautology()
                ? 0
                : BooleanPredicateNormalizer
                        .getDefaultInstanceForCnf()
                        .getMetrics(magicPredicate)
                        .getNormalFormFullSize();
     }
+
+    public static class NormalizedResidualPredicateVisitor implements SimpleExpressionVisitor<QueryPredicate> {
+        @Nonnull
+        @Override
+        public QueryPredicate visitRecordQueryUnionOnValuesPlan(@Nonnull final RecordQueryUnionOnValuesPlan unionPlan) {
```

Comment:
This is unchanged, but this doesn't have a special case for `LogicalUnionExpression`. But I also take it that this predicate may not make sense on logical expressions, in general, as `evaluateAtRef` expects the number of variations in the reference to be one anyway. If we still have some properties that only apply to plans, I could see that suggesting that there is _some_ benefit to a `PlanProperty`--but I'm not sure

Reply:
This is only called from the cost model, thus it only needs to work over references containing exactly one `RecordQueryPlan`.

Reply:
#3322